### PR TITLE
Option to drop temperature after N steps

### PR DESCRIPTION
**Pull request by alejandromarcu**
This is what ChatGPT told me that was used in AlphaZero (not sure if Go or plain, need to double check).
The idea is that after N moves, we drop the temperature for training.  This can be used both in training and in playing.  By using a higher (usually 1) temperature in the beginning, we are more open to exploring (which in play can be useful to start with different openings and not be deterministic), and then when we drop it to 0 we encourage a stronger play.

I'm currently running some experiments.  When using drop_t_on_step=8 it seems to be a bit better than what we currently have on main.  When using drop_t_on_step=16 it gets worse.

image

The command used for drop_t_on_step=8 was:
```
.venv/bin/python -O deep_quoridor/src/train_alphazero.py -N 5 -W 3 \
-g 500 -e 12 --max-steps=200 --num-workers 8 \
--benchmark random \
greedy:p_random=0.5,nick=greedy-05 \
greedy:p_random=0.3,nick=greedy-03 \
greedy:p_random=0.1,nick=greedy-01 \
simple:branching_factor=8,nick=simple-bf8 \
simple:branching_factor=16,nick=simple-bf16 \
-p replay_buffer_size=50000,mcts_ucb_c=2,\
save_replay_buffer=first,mcts_noise_epsilon=0.1,mcts_n=2000,\
learning_rate=0.0001,batch_size=256,optimizer_iterations=100,\
validation_ratio=0.2,drop_t_on_step=8 \
--per-process-evaluation \
-w notes="drop t on step 8"
```

And just changed 8 to 16 for drop_t_on_step=16.   When running in main I just removed that parameter

### CODE REVIEW

File: deep_quoridor/src/agents/alphazero/mcts.py [hunk 1]
```diff
@@ -150,7 +149,7 @@ def select(self, node: Node) -> Node:
 
     def search(self, initial_game: Quoridor):
         root = Node(initial_game, ucb_c=self.ucb_c)
-        self.new_nodes = []
+        new_nodes = []
```
Comment (alejandromarcu) [inline deep_quoridor/src/agents/alphazero/mcts.py:152]: This is unrelated, just piggy-backing a comment by @jonbinney in a previous PR about this var being local

File: deep_quoridor/src/train_alphazero.py [hunk 2]
```diff
@@ -87,6 +87,10 @@ def train_alphazero(
             wandb_train_plugin.episode_count = game_num
             wandb_train_plugin.compute_tournament_metrics(str(current_filename))
 
+    # Close the arena so the best model and the final model are uploaded to wandb
+    if wandb_train_plugin is not None:
+        wandb_train_plugin.end_arena(None, None)
+
```
[the remark this answers is not in the document]
Comment (jonbinney) [reply deep_quoridor/src/train_alphazero.py:93]: Oops! Thanks for fixing this!

**Comment by jonbinney** [comment]
I looked at the papers a bit - the alphazero paper doesn't mention temperature at all, which presumably means they did they same thing as in Alphago Zero.

From the Alphago zero paper:
"For the first 30 moves of each game, the temperature is set to τ = 1; this selects moves proportionally to their visit count in MCTS, andensures a diverse set of positions are encountered. "

AI says "A game of Go does not have a fixed number of moves; games can end at any time if a player resigns, but games finished by counting territory typically last between 200 and 250 moves, though lengths vary, with some finished before 150 moves and others exceeding 350 moves"

So 30 moves of Go is about 30/225 or 13% of one game. Typical Quoridor games are much shorter..... we should do some empirical tests to get a good estimate. But hypothetically if it is about 50 turns on a 5x5x3, then maybe 0.13*50=6.5 turns is a good value. That's pretty close to 8!

File: deep_quoridor/src/agents/alphazero/alphazero.py [hunk 3]
```diff
@@ -228,7 +233,13 @@ def __init__(
...
 
         # When playing use 0.0 for temperature so we always chose the best available action.
-        self.temperature = params.temperature if params.training_mode else 0.0
+        if params.temperature:
+            self.initial_temperature = params.temperature
+        else:
+            if params.training_mode or params.drop_t_on_step is not None:
+                self.initial_temperature = 1
+            else:  # Play mode and params.drop_t_on_step is None
+                self.initial_temperature = 0
```
[the remark this answers is not in the document]
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero/alphazero.py:242]: We were always forcing the temperature to 0 in play mode.  With this change, if you don't specify anything (i.e. None), it will still be 0 in play mode and 1 in training mode, but in either mode you can specify a temperature, and if you also specify drop_t_on_step, it will drop to 0 on that step.

The idea of using temperature in play mode, in conjunction with drop_t, is to give the opportunity to do different openings.  Maybe 1 is too high, but we could use a lower temperature.
For example, the initial move has something like 0.98 score of moving forward, so with temperature 1 it will do it with 98% probability. Then, maybe there are 2 main options that have scores of let's say 0.4 and 0.39.  If temperature is 0, we always choose the one with possibility 0.4, but with that small difference, 0.39 may be as good or better.  If we use temperature 1 we give a chance to both, but also there's a 21% chance that it would do another move that is most likely worse.   But we could give a lower temperature, something like 0.5, so there's a smaller chance of getting a bad move, and the second option still has a fair chance.
And after a certain number of moves, the temperature drops to 0 so we always choose the one we consider the best
